### PR TITLE
NetDb: unit_tests for netdb operations

### PR DESCRIPTION
**Pull request by olarks**
**By submitting this pull-request, I confirm the following:**

- I have read and understood the [Contributing Guide](https://github.com/monero-project/kovri/blob/master/doc/CONTRIBUTING.md).
- I have checked that another pull-request for this purpose does not exist.
- I have considered and confirmed that this submission will be valuable to others.
- I accept that this submission may not be used and that this pull-request may be closed by the will of the maintainer.
- I give this submission freely under the BSD 3-clause license.

*Place an X inside the bracket (or click the box in preview) to confirm*
- [X ] I confirm.

---

All of these tests run. I was going to keep writing tests for netdb requests but I figured I should get some feedback before I move on. I left some TODO's for myself, but the majority of the tests are completed. Some of them I will go over again for completeness, but overall the tests here are all functional.

### CODE REVIEW

**Comment by anonimal** [comment]
@olarks great start. Tests don't pass for reasons specified in the build logs above (and because that's not how one writes unit-tests). I'll give a full review soon.

File: src/core/router/net_db/impl.cc [hunk 1]
```diff
@@ -163,30 +165,14 @@ void NetDb::Run() {
...
-            static_cast<std::uint16_t>(NetDbInterval::DelayedExploratory)) {
-          // Test if we're below the desired threshold
-          if (known_routers <
-              static_cast<std::uint16_t>(NetDbSize::MinKnownRouters)) {
-            // Set the max exploratory count
-            exploratory_count =
-              static_cast<std::uint16_t>(NetDbSize::MaxExploratoryTunnels);
-          }
-        }
+        std::uint16_t exploratory_count = UpdateNumExploratoryTunnels(
```
Comment (anonimal) [inline src/core/router/net_db/impl.cc]: Consider `auto` for type safety / implicit conversion, e.g., if we change the return type for `UpdateNumExploratoryTunnels` to `std::uint32_t` or larger, don't count on remembering to refactor `exploratory_count` unless you have a solid design in place.

File: src/core/router/net_db/impl.h [hunk 2]
```diff
@@ -222,6 +222,12 @@ class NetDb {
       const IdentHash& destination,
       const std::set<IdentHash>& excluded) const;
 
+  bool IsRunning() const {
```
Comment (anonimal) [inline src/core/router/net_db/impl.h]: I don't see the point of this function (it's shouldn't be needed for a unit-test either).

File: src/core/router/net_db/impl.cc [hunk 3]
```diff
@@ -1001,6 +1013,30 @@ std::shared_ptr<const RouterInfo> NetDb::GetClosestNonFloodfill(
...
+  if (known_routers <
+      static_cast<std::uint16_t>(NetDbSize::FavouredKnownRouters)
+      || ts - last_exploratory >=
+      static_cast<std::uint16_t>(NetDbInterval::DelayedExploratory)) {
+    // Test if we're below the desired threshold
+    if (known_routers <
+        static_cast<std::uint16_t>(NetDbSize::MinKnownRouters)) {
+      // Set the max exploratory count
+      exploratory_count =
+        static_cast<std::uint16_t>(NetDbSize::MaxExploratoryTunnels);
```
Comment (anonimal) [inline src/core/router/net_db/impl.cc]: `GetType` here as well throughout the whole function.

File: tests/unit_tests/core/router/net_db/impl.cc [hunk 4]
```diff
@@ -0,0 +1,358 @@
...
+
+#include <boost/test/unit_test.hpp>
+
+#include "core/router/net_db/impl.h"
+#include "core/util/byte_stream.h"
+
+BOOST_AUTO_TEST_SUITE(NetDbTests)
+
+BOOST_AUTO_TEST_CASE(NetDbStart) {
+  using namespace kovri::core;
```
Comment (olarks) [inline tests/unit_tests/core/router/net_db/impl.cc]: Ok. I saw the SSU tests and saw it being scoped like that, so I figured that was how it was done haha.

**Comment by olarks** [comment]
Thanks anonimal. Yeah I squashed the commits because I figured you would have wanted them to get squashed anyway. I guess I'll go back to the drawing board. I appreciate your patience anonimal. I might as well get it right so I can proceed to testing other parts of the code so we aren't just assuming the code work how we think it does.

**Pull request closed without merging**